### PR TITLE
Add metrics and statistics about aggressive locking

### PR DESCRIPTION
**Pull request by MyonKeminta**
Ref: https://github.com/tikv/tikv/issues/13298

### CODE REVIEW

File: metrics/metrics.go [hunk 1]
```diff
@@ -97,6 +97,7 @@ var (
 	TiKVUnsafeDestroyRangeFailuresCounterVec *prometheus.CounterVec
 	TiKVPrewriteAssertionUsageCounter        *prometheus.CounterVec
 	TiKVGrpcConnectionState                  *prometheus.GaugeVec
+	TiKVAggressiveLockedKeysCounter          *prometheus.CounterVec
```
Comment (cfzjywxk) [inline metrics/metrics.go:100]: Do we also need the metric that we could calculate:
- The `aggressive locking keys /  The total locking keys`, so we know how many keys are affected by the locking mode.
- The `aggressive locked transactions/ The total pessimistic transactions`, so we know the percentage affected from transaction perspective.

File: txnkv/transaction/txn.go [hunk 2]
```diff
@@ -1124,13 +1132,32 @@ func (txn *KVTxn) lockKeys(ctx context.Context, lockCtx *tikv.LockCtx, fn func()
...
 				continue
 			}
 			memBuf.UpdateFlags(key, tikv.SetKeyLocked, tikv.DelNeedCheckExists, setValExists)
 		}
 	}
-	txn.lockedCnt += len(keys) - skipedLockKeys
+
+	// Update statistics information.
+	txn.lockedCnt += len(keys) - skippedLockKeys
+	if txn.aggressiveLockingContext != nil && lockCtx.Stats != nil {
```
Comment (cfzjywxk) [inline txnkv/transaction/txn.go:1144]: How about wrapping a statistic or telemetry function.

File: txnkv/transaction/txn.go [hunk 3]
```diff
@@ -988,24 +994,26 @@ func (txn *KVTxn) lockKeys(ctx context.Context, lockCtx *tikv.LockCtx, fn func()
 				return err
 			}
 
+			filteredAggressiveLockedKeysCount = len(allKeys) - len(keys)
+			metrics.AggressiveLockedKeysDerived.Add(float64(filteredAggressiveLockedKeysCount))
+			metrics.AggressiveLockedKeysNew.Add(float64(len(keys)))
+
 			if len(keys) == 0 {
+				if lockCtx.Stats != nil {
+					lockCtx.Stats.AggressiveLockDerivedCount += filteredAggressiveLockedKeysCount
```
Comment (sticnarf) [inline txnkv/transaction/txn.go:1003]: What about moving this out of `if len(keys) == 0`? Then, we make all metrics update and stats update together.

File: metrics/metrics.go (hunk 1 above)
Comment (cfzjywxk) [reply metrics/metrics.go:100]: Oh I see them in the tidb pr.

**Review by sticnarf — COMMENTED** [review]
Only a small optional suggestion.

File: txnkv/transaction/txn.go (hunk 3 above)
Comment (MyonKeminta) [reply txnkv/transaction/txn.go:1003]: Does it matter if `Stats` counts failed ones?

Comment (sticnarf) [reply txnkv/transaction/txn.go:1003]: Ah, seems better not. I see TiDB merges the stats no matter whether there's an error.

File: metrics/shortcuts.go [hunk 4]
```diff
@@ -136,6 +136,11 @@ var (
 	PrewriteAssertionUsageCounterExist    prometheus.Counter
 	PrewriteAssertionUsageCounterNotExist prometheus.Counter
 	PrewriteAssertionUsageCounterUnknown  prometheus.Counter
+
+	AggressiveLockedKeysNew                prometheus.Counter
```
Comment (ekexium) [inline metrics/shortcuts.go:140]: Shall we add brief comments explaining them? For example "derived" is vague if not diving into code.

**Review by ekexium — APPROVED** [review]

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**